[libc] implement quick exit function

aaryanshukla opened the pull request:
Implementing the quick exit function.

aaryanshukla committed:
[libc] implement quick exit function

Implementing the quick exit function.

RoseZhang03 commented on libc/src/stdlib/quick_exit.cpp:
Top and bottom borders (on lines 1 and 7) should be 80 characters exactly

RoseZhang03 commented on libc/src/stdlib/CMakeLists.txt:
Why was -HDRS removed for atof.h?

RoseZhang03 commented on libc/src/__support/libc_assert.h:
Balance the backslash at line 79

RoseZhang03 commented on libc/src/__support/OSUtil/gpu/exit.cpp:
I did LIBC_NAMESPACE::internal, and on the last line I made the comment // namespace LIBC_NAMESPACE::internal, but I think either works

michaelrj-google reviewed (commented):
Here's what I found in review:

aaryanshukla committed:
addressed comments regarding quick exit function

fixed line spacing, max text per line, and unnecesary code that was
moved/deleted.

aaryanshukla committed:
addressing comments to quick exit function

- some files had lines go over 80 pieces of text
- added missing dependencies on quick exit
- reverted some unnecesary changes made to the make files
- fixed the order of include files to met clang formatting rules.

RoseZhang03 reviewed (commented):
LGTM!

michaelrj-google commented on libc/src/stdlib/quick_exit.cpp:
```diff
@@ -10,16 +10,13 @@
- // namespace LIBC_NAMESPACE
-
+} // namespace LIBC_NAMESPACE
+  // namespace LIBC_NAMESPACE
```
nit: Remove extra `LIBC_NAMESPACE` comment

nickdesaulniers commented on libc/src/stdlib/quick_exit.cpp:
```diff
@@ -0,0 +1,22 @@
+#include "src/__support/OSUtil/exit.h"
+#include "src/__support/common.h"
+
+// extern "C" void __cxa_finalize(void *);
+namespace LIBC_NAMESPACE {
+
+[[noreturn]] LLVM_LIBC_FUNCTION(void, quick_exit, (int status)) {
+  //__cxa_finalize(nullptr);
```
What's up with these comments?

nickdesaulniers commented on libc/src/stdlib/quick_exit.h:
```diff
@@ -0,0 +1,20 @@
+#ifndef LLVM_LIBC_SRC_STDLIB_QUICK_EXIT_H
+#define LLVM_LIBC_SRC_STDLIB_QUICK_EXIT_H
+
+#include <stdlib.h>
```
Do we need anything from stdlib to satisfy this declaration? I don't think so. Can you remove it?

nickdesaulniers commented on libc/test/IntegrationTest/test.h:
```diff
@@ -9,40 +9,41 @@
-    if (should_exit)                                                           \
-      LIBC_NAMESPACE::quick_exit(127);                                         \
-  }
+    if (should_exit)
+
+LIBC_NAMESPACE::internal::exit(127);
```
formatting looks wrong here.

nickdesaulniers commented on libc/test/src/stdlib/quick_exit_test.cpp:
```diff
@@ -0,0 +1,17 @@
+#include "src/stdlib/quick_exit.h"
+#include "test/UnitTest/Test.h"
+
+#include <stdlib.h>
```
do we need anything from stdlib here? I don't think so; can you remove it?

lntue commented on libc/src/__support/OSUtil/baremetal/exit.cpp:
```diff
@@ -1,18 +1,21 @@
-//===----- Baremetal implementation of a quick exit function ----*- C++ -*-===//
+//===----- Baremetal implementation of a exit function ----*- C++ -*-===//
```
Nit: alignment

lntue commented on libc/src/stdlib/CMakeLists.txt:
```diff
@@ -432,6 +432,18 @@ add_entrypoint_object(
+  HDRS
+    quick_exit.h
+  DEPENDS
+    libc.src.errno.errno
```
Does not look like `errno` is used in the implementation?

lntue commented on libc/src/__support/OSUtil/linux/exit.cpp:
```diff
@@ -10,19 +10,18 @@
 #ifdef LIBC_TARGET_ARCH_IS_X86
 __attribute__((no_stack_protector))
 #endif
 __attribute__((noreturn)) void
```
This one probably should be changed to `[[noreturn]]` to match others.

nickdesaulniers replied:
I think there was an issue with `[[]]` notation for multiple attributes, IIRC.

aaryanshukla replied:
keeping ((noreturn)) the same for now.

jhuber6 replied:
https://godbolt.org/z/rhYeP3vjM which part doesn't work?

nickdesaulniers replied:
ah, then perhaps I misremember. Sorry for not just checking it myself!

aaryanshukla replied:
should this just be attribute[[noreturn]] then?

nickdesaulniers replied:
ah, GCC doesn't recognize the `[[clang::` prefix, that was the issue! Build is now broken from https://github.com/llvm/llvm-project/pull/93620

lntue replied:
using `[[gnu::` instead should work for gcc: https://godbolt.org/z/6dbrc6cY6

RoseZhang03 replied:
Clang doesn't recognize [[gnu::

lntue replied:
hmm, interesting, turns out it is supported since clang 18: https://godbolt.org/z/G1srfsWen

SchrodingerZhu replied:
I should have put a comment there. 🤣

This is the third time ppl get confused.

SchrodingerZhu replied:
> hmm, interesting, turns out it is supported since clang 18: https://godbolt.org/z/G1srfsWen

Because we filed a issue just for this place.

aaryanshukla committed:
fixed no_stack_proctector brackets and noreturn, including formatting

added no_stack_proctector and noreturn brackets.
removed unnecesary stdlib.h files.
fixed line formatting for test.h files.

aaryanshukla committed:
added internal in front of exit for quick_exit

put internal in front of the calling for exit.

aaryanshukla committed:
updated build.BAZEL, changed all quick_exit to exit

changed all areas from quick_exit to exit in build.BAZEL.

aaryanshukla committed:
changed quick_exit to exit in cmakefiles in gpu and baremetal

modified quick_exit to exit in cmakefiles.

michaelrj-google reviewed (approved):
LGTM

lntue commented on libc/src/__support/OSUtil/baremetal/exit.cpp:
```diff
@@ -1,18 +1,21 @@
-//===----- Baremetal implementation of a quick exit function ----*- C++ -*-===//
+//===----- Baremetal implementation of an exit function ----*- C++ -*-===//
```
nit: align the length of this line.

michaelrj-google commented:
Superseded by #93620
